File: evals/eval_math_agent.py
"""Math Agent evaluation focused on calculation accuracy and tool usage."""

import os
import re
import sys
from pathlib import Path
from typing import Any

# Ensure project root is on sys.path
project_root = Path(__file__).resolve().parents[1]
if str(project_root) not in sys.path:
    sys.path.insert(0, str(project_root))

from autoevals import LLMClassifier  # noqa: E402
from braintrust import Eval  # noqa: E402
from dotenv import load_dotenv  # noqa: E402

from evals.parameters import MathAgentPromptParam, MathModelParam  # noqa: E402
from src.agents.math_agent import get_math_agent  # noqa: E402
from src.helpers import run_adk_agent  # noqa: E402
from src.tracing import configure_adk_tracing  # noqa: E402

load_dotenv()
configure_adk_tracing(
    api_key=os.environ.get("BRAINTRUST_API_KEY"),
    project_id=os.environ.get("BRAINTRUST_PROJECT_ID"),
    project_name=os.environ.get("BRAINTRUST_PROJECT", "vercel-ai-sdk-supervisor"),
)


def _param_value(param: Any, default: Any) -> Any:
    if param is None:
        return default
    if hasattr(param, "value"):
        return getattr(param, "value")
    if isinstance(param, type):
        try:
            instance = param()
            if hasattr(instance, "value"):
                return getattr(instance, "value")
        except Exception:
            return default
    return param


async def run_math_task(input: dict, hooks: Any = None) -> dict:
    """Run a math calculation through the math agent."""
    try:
        params = hooks.parameters if hooks and hasattr(hooks, "parameters") else {}
        math_agent_prompt = _param_value(params.get("math_agent_prompt"), None)
        math_model = _param_value(params.get("math_model"), "gemini-2.0-flash-lite")

        agent = get_math_agent(system_prompt=math_agent_prompt, model=math_model)
        query = str(input.get("query", ""))

        run_result = await run_adk_agent(
            agent=agent,
            query=query,
            app_name="google-adk-supervisor-eval-math",
        )
        serialized = run_result["messages"]

        tool_calls: list[dict[str, Any]] = []
        for msg in serialized:
            for tc in msg.get("tool_calls", []):
                tool_calls.append(
                    {
                        "name": tc.get("name", ""),
                        "args": tc.get("args", {}),
                    }
                )

        if hooks and hasattr(hooks, "metadata"):
            hooks.metadata.update(
                {
                    "tool_calls": tool_calls,
                    "total_messages": len(serialized),
                }
            )

        return {"messages": serialized}
    except Exception as e:
        if hooks and hasattr(hooks, "metadata"):
            hooks.metadata.update({"error": str(e)})
        return {"messages": [{"error": str(e)}]}


MATH_TEST_DATA = [
    {"input": {"query": "What is 25 + 17?", "expected_answer": 42}},
    {"input": {"query": "Calculate 100 - 37", "expected_answer": 63}},
    {"input": {"query": "What is 12 * 8?", "expected_answer": 96}},
    {"input": {"query": "Divide 144 by 12", "expected_answer": 12}},
    {"input": {"query": "What's 15 * 7 + 3?", "expected_answer": 108}},
    {"input": {"query": "Calculate (50 + 30) / 4", "expected_answer": 20}},
]


async def calculation_accuracy_scorer(input, output, expected):
    """Check if the calculated answer matches the expected value."""
    if not expected or "expected_answer" not in expected:
        return 0.5

    expected_answer = expected["expected_answer"]
    messages = output.get("messages", [])

    for msg in reversed(messages):
        content = msg.get("content", "") if isinstance(msg, dict) else ""
        role = msg.get("role", "") if isinstance(msg, dict) else ""
        if content and role == "assistant":
            if str(expected_answer) in str(content):
                return 1.0
            break

    return 0.0


async def tool_usage_scorer(output, metadata=None):
    """Check if the agent used math tools appropriately."""
    if not metadata:
        return 0.5

    tool_calls = metadata.get("tool_calls", [])
    if not tool_calls:
        return 0.0

    tool_names = [tc["name"] for tc in tool_calls]
    valid_tools = {"add", "subtract", "multiply", "divide"}
    used_valid_tools = any(name in valid_tools for name in tool_names)
    return 1.0 if used_valid_tools else 0.0


async def efficiency_scorer(output, metadata=None):
    """Score based on minimal unnecessary tool calls."""
    if not metadata:
        return 0.5

    num_calls = len(metadata.get("tool_calls", []))
    if num_calls <= 2:
        return 1.0
    if num_calls <= 4:
        return 0.8
    if num_calls <= 6:
        return 0.6
    return 0.4


async def response_format_scorer(output):
    """Check if the response is clear and includes the final answer."""
    messages = output.get("messages", [])
    for msg in reversed(messages):
        content = msg.get("content", "") if isinstance(msg, dict) else ""
        role = msg.get("role", "") if isinstance(msg, dict) else ""

        if content and role == "assistant":
            if re.search(r"\d+", content):
                return 1.0
            break
    return 0.0


calculation_correctness_prompt = """
You are evaluating a math agent's calculation.

Question: {{input}}
Agent's Response: {{output}}
Expected Answer: {{expected}}

Evaluate whether:
1. The calculation is mathematically correct
2. The final answer matches the expected result
3. The reasoning (if shown) is sound

Respond with:
CORRECT - Calculation and answer are correct
INCORRECT - Calculation or answer is wrong
"""

calculation_correctness_scorer = LLMClassifier(
    name="Calculation Correctness",
    prompt_template=calculation_correctness_prompt,
    choice_scores={"CORRECT": 1.0, "INCORRECT": 0.0},
    use_cot=True,
    model=os.environ.get("EVAL_JUDGE_MODEL", "gpt-4o"),
)


Eval(
    os.environ.get("BRAINTRUST_PROJECT", "vercel-ai-sdk-supervisor"),
    experiment_name="math-agent",
    data=MATH_TEST_DATA,  # type: ignore
    task=run_math_task,
    scores=[
        calculation_accuracy_scorer,
        tool_usage_scorer,
        efficiency_scorer,
        response_format_scorer,
        calculation_correctness_scorer,
    ],  # type: ignore
    parameters={
        "math_agent_prompt": MathAgentPromptParam,
        "math_model": MathModelParam,
    },
)

File: evals/eval_math_agent.ts
import "dotenv/config";

import { Eval } from "braintrust";
import { z } from "zod";

import { DEFAULT_MATH_AGENT_PROMPT } from "../src-ts/config.js";
import { runMathAgent } from "../src-ts/agents/math-agent.js";
import { extractFloatFromText } from "../src-ts/serializer.js";
import { configureTracing } from "../src-ts/tracing.js";
import type { SerializedMessage } from "../src-ts/types.js";

configureTracing({
  apiKey: process.env.BRAINTRUST_API_KEY,
  projectName: process.env.BRAINTRUST_PROJECT ?? "vercel-ai-sdk-supervisor",
  projectId: process.env.BRAINTRUST_PROJECT_ID,
});

const parameters = {
  math_agent_prompt: z.string().default(DEFAULT_MATH_AGENT_PROMPT),
  math_model: z.string().default("gpt-4.1-mini"),
};

type MathTaskInput = { query: string; expected_answer?: number };
type MathTaskOutput = { messages: SerializedMessage[] };

async function runMathTask(input: MathTaskInput, hooks: { parameters: { math_agent_prompt: string; math_model: string }; metadata: Record<string, unknown> }): Promise<MathTaskOutput> {
  try {
    const result = await runMathAgent({
      query: input.query,
      systemPrompt: hooks.parameters.math_agent_prompt,
      model: hooks.parameters.math_model,
    });

    const toolCalls = result.messages.flatMap((message) => message.tool_calls ?? []);
    hooks.metadata.tool_calls = toolCalls;
    hooks.metadata.total_messages = result.messages.length;

    return { messages: result.messages };
  } catch (error) {
    hooks.metadata.error = String(error);
    return { messages: [{ role: "system", content: `error: ${String(error)}` }] };
  }
}

const MATH_TEST_DATA: Array<{
  input: MathTaskInput;
  expected: { expected_answer: number };
  metadata: Record<string, unknown>;
}> = [
  { input: { query: "What is 25 + 17?" }, expected: { expected_answer: 42 }, metadata: {} },
  { input: { query: "Calculate 100 - 37" }, expected: { expected_answer: 63 }, metadata: {} },
  { input: { query: "What is 12 * 8?" }, expected: { expected_answer: 96 }, metadata: {} },
  { input: { query: "Divide 144 by 12" }, expected: { expected_answer: 12 }, metadata: {} },
  { input: { query: "What's 15 * 7 + 3?" }, expected: { expected_answer: 108 }, metadata: {} },
  { input: { query: "Calculate (50 + 30) / 4" }, expected: { expected_answer: 20 }, metadata: {} },
];

function latestAssistant(messages: SerializedMessage[]): string {
  for (let index = messages.length - 1; index >= 0; index -= 1) {
    const message = messages[index];
    if (message.role === "assistant" && message.content.trim()) {
      return message.content;
    }
  }
  return "";
}

async function calculationAccuracyScorer(args: {
  output: MathTaskOutput;
  expected: { expected_answer: number };
}) {
  const answerText = latestAssistant(args.output.messages);
  const parsed = extractFloatFromText(answerText);
  const expected = args.expected.expected_answer;

  if (parsed === null) {
    return { name: "Calculation Accuracy", score: 0 };
  }

  const score = Math.abs(parsed - expected) < 1e-6 ? 1 : 0;
  return { name: "Calculation Accuracy", score };
}

async function toolUsageScorer(args: { output: MathTaskOutput }) {
  const toolNames = args.output.messages
    .flatMap((message) => message.tool_calls ?? [])
    .map((toolCall) => toolCall.name);

  const validTools = new Set(["add", "subtract", "multiply", "divide"]);
  const usedValidTools = toolNames.some((name) => validTools.has(name));
  return { name: "Tool Usage", score: usedValidTools ? 1 : 0 };
}

async function efficiencyScorer(args: { output: MathTaskOutput }) {
  const numCalls = args.output.messages.flatMap((message) => message.tool_calls ?? []).length;
  if (numCalls <= 2) {
    return { name: "Efficiency", score: 1 };
  }
  if (numCalls <= 4) {
    return { name: "Efficiency", score: 0.8 };
  }
  if (numCalls <= 6) {
    return { name: "Efficiency", score: 0.6 };
  }
  return { name: "Efficiency", score: 0.4 };
}

async function responseFormatScorer(args: { output: MathTaskOutput }) {
  const answerText = latestAssistant(args.output.messages);
  return { name: "Response Format", score: /\d/.test(answerText) ? 1 : 0 };
}

await Eval(process.env.BRAINTRUST_PROJECT ?? "vercel-ai-sdk-supervisor", {
  experimentName: "math-agent",
  data: MATH_TEST_DATA,
  task: runMathTask,
  scores: [calculationAccuracyScorer, toolUsageScorer, efficiencyScorer, responseFormatScorer],
  parameters,
});
